persistence/remote/src/main/java/org/infinispan/persistence/remote/upgrade/MigrationTask.java

```diff
@@ -37,46 +51,97 @@
    private final Set<Integer> segments;
    private final int readBatch;
    private final int threads;
-   private volatile Cache<Object, Object> cache;
+   private final ConcurrentHashSet<WrappedByteArray> deletedKeys = new ConcurrentHashSet<>();
+   private byte[] ignoredKey;
+
+   private transient Set<RemoteStore> stores;
+   private transient Cache<Object, Object> cache;
+   private transient ExecutorService executorService;
+   private transient RemoveListener listener = new RemoveListener();
 
    public MigrationTask(Set<Integer> segments, int readBatch, int threads) {
       this.segments = segments;
       this.readBatch = readBatch;
       this.threads = threads;
    }
 
+   @Listener(clustered = true)
+   @SuppressWarnings("unused")
+   private class RemoveListener {
+      @CacheEntryRemoved
+      public void entryRemoved(CacheEntryRemovedEvent event) {
+         deletedKeys.add(new WrappedByteArray((byte[]) event.getKey()));
+      }
+   }
+
    @Override
    public Integer call() throws Exception {
-      ComponentRegistry cr = cache.getAdvancedCache().getComponentRegistry();
-      PersistenceManager loaderManager = cr.getComponent(PersistenceManager.class);
-      Set<RemoteStore> stores = loaderManager.getStores(RemoteStore.class);
-      Marshaller marshaller = new MigrationMarshaller();
-      byte[] knownKeys;
       try {
-         knownKeys = marshaller.objectToByteBuffer(MIGRATION_MANAGER_HOT_ROD_KNOWN_KEYS);
-      } catch (Exception e) {
-         throw new CacheException(e);
+         Iterator<RemoteStore> storeIterator = stores.iterator();
```

Comment:
In theory yes, but I am a bit unsure if having a cache with more than 1 remote store, each one pointing to a different source cluster potentially, being part of a N-1 migration is something useful...